pkg/services/ngalert/eval/alert_query.go

```diff
@@ -0,0 +1,289 @@
+	if err != nil {
+		return 0, err
+	}
+
+	maxDataPoints, ok := aq.modelProps["maxDataPoints"].(float64)
+	if !ok {
+		return 0, fmt.Errorf("failed to cast maxDataPoints to float64: %v", aq.modelProps["maxDataPoints"])
+	}
+	return int64(maxDataPoints), nil
+}
+
+// setIntervalMS sets the model IntervalMs if it's missing or invalid
+func (aq *AlertQuery) setIntervalMS() error {
+	if aq.modelProps == nil {
+		err := aq.setModelProps()
+		if err != nil {
+			return err
+		}
+	}
+	i, ok := aq.modelProps["intervalMs"] // GEL requires intervalMs inside the query JSON
+	if !ok {
+		aq.modelProps["intervalMs"] = defaultIntervalMS
+	}
+	intervalMs, ok := i.(float64)
+	if !ok || intervalMs == 0 {
+		aq.modelProps["intervalMs"] = defaultIntervalMS
+	}
+	return nil
+}
+
+func (aq *AlertQuery) getIntervalMS() (int64, error) {
+	err := aq.setIntervalMS()
+	if err != nil {
+		return 0, err
+	}
+
+	intervalMs, ok := aq.modelProps["intervalMs"].(float64)
+	if !ok {
+		return 0, fmt.Errorf("failed to cast intervalMs to float64: %v", aq.modelProps["intervalMs"])
+	}
+	return int64(intervalMs), nil
+}
+
+// GetDatasource returns the query datasource identifier.
+func (aq *AlertQuery) GetDatasource() (int64, error) {
+	err := aq.setDatasource()
+	if err != nil {
+		return 0, err
+	}
+	return aq.DatasourceID, nil
+}
+
+func (aq *AlertQuery) getModel() ([]byte, error) {
+	err := aq.setDatasource()
+	if err != nil {
+		return nil, err
+	}
+
+	err = aq.setMaxDatapoints()
+	if err != nil {
+		return nil, err
+	}
+
+	err = aq.setIntervalMS()
+	if err != nil {
+		return nil, err
+	}
+
+	model, err := json.Marshal(aq.modelProps)
+	if err != nil {
+		return nil, fmt.Errorf("unable to marshal query model: %w", err)
+	}
+	return model, nil
+}
+
+func (aq *AlertQuery) setOrgID(orgID int64) error {
+	if aq.modelProps == nil {
+		err := aq.setModelProps()
+		if err != nil {
+			return err
+		}
+	}
+
+	aq.modelProps["orgId"] = orgID
+	return nil
+}
+
+func (aq *AlertQuery) setQueryType() error {
+	if aq.modelProps == nil {
+		err := aq.setModelProps()
+		if err != nil {
+			return err
+		}
+	}
+	i, ok := aq.modelProps["queryType"]
+	if !ok {
+		return nil
+	}
+
+	queryType, ok := i.(string)
+	if !ok {
+		return fmt.Errorf("failed to get queryType from query model: %v", i)
+	}
+
+	aq.QueryType = queryType
+	return nil
+}
+
+// PreSave sets queries properties.
+// It should be called before being saved.
+func (aq *AlertQuery) PreSave(orgID int64) error {
+	err := aq.setOrgID(orgID)
+	if err != nil {
+		return fmt.Errorf("failed to set orgId to query model: %w", err)
+	}
+
+	if err := aq.setDatasource(); err != nil {
+		return fmt.Errorf("failed to set datasource to query model: %w", err)
+	}
+
+	if err := aq.setQueryType(); err != nil {
+		return fmt.Errorf("failed to set query type to query model: %w", err)
+	}
+
+	// override model
+	model, err := aq.getModel()
+	if err != nil {
+		return err
+	}
+	aq.Model = model
+
+	if ok := aq.RelativeTimeRange.isValid(); !ok {
```

Comment:
Currently backend expression nodes don't have a time range. They might, but if they do, it might be as parameters to whatever is being called. So for now a valid time range is not needed expected for the expression queries, just datasource ones.